### DISCUSSION TITLE
Newbie q: if I create a ROCm/HIP plugin (.dll) in AMD mode, does it use standard OpenCL?

### DISCUSSION THREAD

**Discussion by garyo** [ROCm/ROCm#3353] [Q&A]
I'm creating a plugin (.dll on Windows) and the calling app is going to pass me OpenCL buffers and images. If I use HIP to create my plugin, can I build it for Windows so it is using OpenCL, and will accept those images and buffers? The docs don't seem super clear on this.

**Comment by fluidnumericsJoe** [comment]
I wouldn't use HIP for handling OpenCL buffers. ROCm's LLVM-based amdclang compiler works with OpenCL. You should just be able to work with OpenCL directly in your app.

**Comment by garyo** [comment]
Thanks for this, Joe -- I've used OpenCL before, but not in quite a few
years. Are you suggesting ROCm as an alternative compiler for writing
OpenCL (vs. the Khronos SDK)? I'd love to learn more -- anything to make my
OpenCL implementation easier and better would be very appreciated.

On Wed, Jun 26, 2024 at 4:25 PM Joe Schoonover ***@***.***>
wrote:

> I wouldn't use HIP for handling OpenCL buffers. ROCm's LLVM-based amdclang
> compiler works with OpenCL. You should just be able to work with OpenCL
> directly in your app.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ROCm/ROCm/discussions/3353#discussioncomment-9887048>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABCFR5WXPYV7XDDMGGNMF3ZJMPTHAVCNFSM6AAAAABJZZTA5SVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TQOBXGA2DQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Gary Oberbrunner
Founder & CEO
BioViz Studio <https://www.bioviz-studio.com> from Dark Star Systems, Inc
<https://darkstarsystems.com>.
[image: LinkedIn] <https://www.linkedin.com/company/darkstarsystems>
<https://twitter.com/darkstarsystems>
<https://instagram.com/darkstarsystems>
<https://www.facebook.com/darkstarsys>

**Reply by fluidnumericsJoe** [reply]
I suppose the compiler choice is up to you. If you're looking to use ROCm specifically, ROCm ships with an llvm-based compiler suite (amdclang, amdclang++, amdflang) that are capable of compiling opencl kernels in addition to opencl headers. Sharing memory between OpenCL and HIP (and OpenCL and CUDA) is a tough road to go down in my experience.

**Comment by MathiasMagnus** [comment]
@garyo The short answer is "no". The little longer one "in an ideal world yes, in the one we live in, no". (In the ideal world, your plug-in probably would still end up depending on a HIP runtime, so technically it would cease being a strictly speaking OpenCL-only app.)

In the coming part, ICD (Installable Client Driver), runtime and UMD (User-Mode Driver) are used interchangably.

HIP and OpenCL are two separate APIs. ROCm under the hood is using a common runtime for both (CLR, aka. Common Language Runtime) on top of which thin wrappers provide the HIP (in AMD mode) and OpenCL interfaces. That means that the two should trivially interop (much like OpenCL-OpenGL interop, where you can map API handles between the APIs), but those interop APIs aren't implemented.

When you're developing a plug-in, you'll likely be writing both host-side code and device-side code as well, you'll want to enqueue kernels on the GPU, kernels you have to build. In OpenCL, you use regular OpenCL calls for both, the runtime has your vendor's compiler inside it. HIP is different, HIP is typically single-source (though it has a run-time compiler too), but if you use any HIP host-side APIs to compile code or to dispatch data (using the non-existing interop), you'll be going through the HIP runtime, at which point in AMD-mode you'll depend on amdhip64.dll. At this point it'll only work on AMD and stop being a portable OpenCL plugin.

The beauty of OpenCL is that you need no special compilers as far as the host side goes to compile it (and as @fluidnumerics-joe pointed out, AMDClang shipping with the HIP SDK is just as good as MSVC or GCC), and the ICDs shipping with drivers have the device compiler inside them.

_(For the sake of understanding, AMD's OpenCL ICD has [comgr](https://github.com/ROCm/llvm-project/tree/amd-staging/amd/comgr) (Code Object Manager) inside it, which is a nicer/simpler API to use LLVM as a library, as opposed to LLVM's own interfaces. The device-code produced by matching driver/SDK versions should be identical, whether you feed your OpenCL kernel to AMD's ICD and [`clGetProgramInfo(CL_PROGRAM_BINARIES)`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clGetProgramInfo) or you compile them offline using amdclang and feed them to AMD's ICD using [`clCreateProgramWithBinary`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clCreateProgramWithBinary))_

You only need the headers (and an import lib on Windows) to link against, even the DLL part (the ICD loader) is shipped with vendor drivers. The HIP SDK for Windows is (_nomen est omen_) a HIP SDK, there are no OpenCL headers or other dev components shipping in it. For that the Khronos SDK is probably your best bet, though again, you can use any vendor's OpenCL SDK to build OpenCL apps, it will work everywhere. AMD used to have OpenCL dev files shipping in the APP SDK days, but those days are gone. While ROCm runtimes do ship OpenCL drivers, the ROCm SDK don't ship OpenCL headers. (AMD would want you to write HIP, not OpenCL.)
